access/grpc/grpc.go

```diff
@@ -1296,3 +1298,159 @@ func receiveBlocksFromClient[Client interface {
+	request := &executiondata.SubscribeAccountStatusesFromStartBlockIDRequest{
+		StartBlockId:         startBlockID.Bytes(),
+		EventEncodingVersion: c.eventEncoding,
+	}
+	request.Filter = &executiondata.StatusFilter{
+		EventType: filter.EventTypes,
+		Address:   filter.Addresses,
+	}
+
+	subscribeClient, err := c.executionDataClient.SubscribeAccountStatusesFromStartBlockID(ctx, request, opts...)
+	if err != nil {
+		return nil, nil, newRPCError(err)
+	}
+
+	accountStatutesChan := make(chan flow.AccountStatus)
+	errChan := make(chan error)
+
+	go func() {
+		defer close(accountStatutesChan)
+		defer close(errChan)
+		receiveAccountStatusesFromClient(ctx, subscribeClient, accountStatutesChan, errChan, &c.messageIndex)
+	}()
+
+	return accountStatutesChan, errChan, nil
+}
+
+func (c *BaseClient) SubscribeAccountStatusesFromLatestBlock(
+	ctx context.Context,
+	filter flow.AccountStatusFilter,
+	opts ...grpc.CallOption,
+) (<-chan flow.AccountStatus, <-chan error, error) {
+	request := &executiondata.SubscribeAccountStatusesFromLatestBlockRequest{
+		EventEncodingVersion: c.eventEncoding,
+	}
+	request.Filter = &executiondata.StatusFilter{
+		EventType: filter.EventTypes,
+		Address:   filter.Addresses,
+	}
+
+	subscribeClient, err := c.executionDataClient.SubscribeAccountStatusesFromLatestBlock(ctx, request, opts...)
+	if err != nil {
+		return nil, nil, newRPCError(err)
+	}
+
+	accountStatutesChan := make(chan flow.AccountStatus)
+	errChan := make(chan error)
+
+	go func() {
+		defer close(accountStatutesChan)
+		defer close(errChan)
+		receiveAccountStatusesFromClient(ctx, subscribeClient, accountStatutesChan, errChan, &c.messageIndex)
+	}()
+
+	return accountStatutesChan, errChan, nil
+}
+
+func receiveAccountStatusesFromClient[Client interface {
+	Recv() (*executiondata.SubscribeAccountStatusesResponse, error)
+}](
+	ctx context.Context,
+	client Client,
+	accountStatutesChan chan<- flow.AccountStatus,
+	errChan chan<- error,
+	previousMsgIndex *atomic.Uint64,
+) {
+	sendErr := func(err error) {
+		select {
+		case <-ctx.Done():
+		case errChan <- err:
+		}
+	}
+
+	for {
+		accountStatusResponse, err := client.Recv()
+		if err != nil {
+			if err == io.EOF {
+				// End of stream, return gracefully
+				return
+			}
+
+			sendErr(fmt.Errorf("error receiving account status: %w", err))
+			return
+		}
+
+		accountStatus, err := convert.MessageToAccountStatus(accountStatusResponse)
+		if err != nil {
+			sendErr(fmt.Errorf("error converting message to account status: %w", err))
+			return
+		}
+
+		if err = checkAndIncrementMessageIndex(previousMsgIndex, accountStatus.MessageIndex); err != nil {
+			sendErr(fmt.Errorf("error checking message index. messages are not ordered: %w", err))
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case accountStatutesChan <- accountStatus:
+		}
+	}
+}
+
+func checkAndIncrementMessageIndex(previousMessageIndex *atomic.Uint64, currentMessageIndex uint64) error {
```

Comment:
@peterargue is this implementation good enough? It might be good to add more tests for different msg indexes. What do you think?

Reply:
I think you can simply it a bit by tracking "next" instead of the current.
e.g.
```golang
next := 0 // since we start at 0, the first expected should be 0
for {
	if msg.MessageIndex != next {
		return err
	}
	next = msg.MessageIndex + 1
}
```